### PR TITLE
Add RP2 PIO background read

### PR DESCRIPTION
**Pull request by timchinowsky**
This PR adds a `background_read` method to `rp2pio.StateMachine` to complement the existing `background_write`.

### CODE REVIEW

File: ports/raspberrypi/bindings/rp2pio/StateMachine.c [hunk 1]
```diff
@@ -618,23 +619,166 @@ const mp_obj_property_t rp2pio_statemachine_writing_obj = {
...
+            mp_raise_ValueError(MP_ERROR_TEXT("Buffer elements must be 4 bytes long or less"));
+        }
+        if (*stride_in_bytes && stride != *stride_in_bytes) {
+            mp_raise_ValueError(MP_ERROR_TEXT("Mismatched data size"));
+        }
+        *stride_in_bytes = stride;
+    } else {
+        memset(info, 0, sizeof(*info));
+    }
+}
```
Comment (jepler) [inline ports/raspberrypi/bindings/rp2pio/StateMachine.c:695]: Can you please combine the fill_buf_info_{read,write} routines? I notice you've added detection of mismatched stride between the "once" and "loop" variants, which should probably apply to write as well. I think you'd just have to add the access type parameter and pass MP_BUFFER_{READ,WRITE} in.

Or is there a reason the new checking you added should not also have been done for the background write case?

File: ports/raspberrypi/bindings/rp2pio/StateMachine.c [hunk 2]
```diff
@@ -618,23 +619,166 @@ const mp_obj_property_t rp2pio_statemachine_writing_obj = {
 };
 
 
+
 //|     pending: int
+//|     pending_write: int
```
Comment (jepler) [inline ports/raspberrypi/bindings/rp2pio/StateMachine.c:624]: Please document that the `pending` property is a deprecated alias for `pending_write` and will be removed in a future version of CircuitPython.

File: ports/raspberrypi/bindings/rp2pio/StateMachine.c [hunk 3]
```diff
@@ -618,23 +619,166 @@ const mp_obj_property_t rp2pio_statemachine_writing_obj = {
...
 //|
 //|     If the number is 0, then a `StateMachine.background_write` call will not block."""
-static mp_obj_t rp2pio_statemachine_obj_get_pending(mp_obj_t self_in) {
+static mp_obj_t rp2pio_statemachine_obj_get_pending_write(mp_obj_t self_in) {
+    rp2pio_statemachine_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    return mp_obj_new_int(common_hal_rp2pio_statemachine_get_pending_write(self));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(rp2pio_statemachine_get_pending_write_obj, rp2pio_statemachine_obj_get_pending_write);
+
+const mp_obj_property_t rp2pio_statemachine_pending_write_obj = {
```
Comment (jepler) [inline ports/raspberrypi/bindings/rp2pio/StateMachine.c:634]: Please use the MP_PROPERTY_GETTER macro here, even though it wasn't used in the code whose style you were emultaing elsewhere in this file.  We should tidy away the direct uses of `const mp_obj_property_t` that have not previously been done, like the other ones in this file.

File: ports/raspberrypi/common-hal/rp2pio/StateMachine.c [hunk 4]
```diff
@@ -1023,121 +1047,241 @@ uint8_t rp2pio_statemachine_program_offset(rp2pio_statemachine_obj_t *self) {
     return _current_program_offset[pio_index][sm];
 }
 
-bool common_hal_rp2pio_statemachine_background_write(rp2pio_statemachine_obj_t *self, const sm_buf_info *once, const sm_buf_info *loop, uint8_t stride_in_bytes, bool swap) {
+bool common_hal_rp2pio_statemachine_background_write(rp2pio_statemachine_obj_t *self, const sm_buf_info *once_write, const sm_buf_info *loop_write, uint8_t stride_in_bytes, bool swap) {
```
Comment (jepler) [inline ports/raspberrypi/common-hal/rp2pio/StateMachine.c:1050]: I wouldn't personally have changed the parameter and local variables names but this is OK.

File: ports/raspberrypi/common-hal/rp2pio/StateMachine.h [hunk 5]
```diff
@@ -47,10 +47,13 @@ typedef struct {
...
     // dma-related items
-    volatile int pending_buffers;
-    sm_buf_info current, once, loop;
+    volatile int pending_buffers_write;
+    volatile int pending_buffers_read;
+    sm_buf_info current_write, once_write, loop_write;
+    sm_buf_info current_read, once_read, loop_read;
     int background_stride_in_bytes;
-    bool dma_completed, byteswap;
+    bool dma_completed_write, byteswap;
```
Comment (jepler) [inline ports/raspberrypi/common-hal/rp2pio/StateMachine.h:55]: hmm  are two byteswap values (one for read, one for write) needed?

**Review by jepler — CHANGES_REQUESTED** [review]
Thanks for this, I'm excited!

I did have some review comments & requested changes. I also didn't actually test the code. Any suggestion on something that would be useful to try/test if I don't have a TAC5xxx codec?

**Comment by timchinowsky** [comment]
> Thanks for this, I'm excited!
> 
> I did have some review comments & requested changes. I also didn't actually test the code. Any suggestion on something that would be useful to try/test if I don't have a TAC5xxx codec?

For testing with no codec, I'm getting together a loopback example similar to https://github.com/adafruit/Adafruit_CircuitPython_PIOASM/blob/main/examples/pioasm_i2s_codec.py that will send a `background_write` to a `background_read` by connecting output and input data pins.  If you're interested in messing with the codec itself, I've got some extras I could send you, let me know.

**Comment by timchinowsky** [comment]
> Any suggestion on something that would be useful to try/test if I don't have a TAC5xxx codec?

I've added a loopback example at https://github.com/timchinowsky/tac5 which you should be able to run as is on a Feather RP2040 if you connect DOUT (D9) to DIN (D10).

File: ports/raspberrypi/common-hal/rp2pio/StateMachine.c (hunk 4 above)
Comment (timchinowsky) [reply ports/raspberrypi/common-hal/rp2pio/StateMachine.c]: I wanted to make it clearer that the parameters are buffers not flags, as `loop` is used for instance in https://github.com/adafruit/circuitpython/blob/main/ports/raspberrypi/audio_dma.c

**Comment by timchinowsky** [comment]
The last two commits add double buffering and a `last_read` property which (a) returns the buffer which was last completely filled by background writes and (b) self-clears, so that if the property is read again before the next buffer swap, it will return an empty buffer.  This allows data read in the background to be recorded to SD with code as simple as

```python
with open(filename, 'w') as f:
    while True:
        f.write(self.pcm.pio.last_read)
```

With the double-buffering and `last_read` property there are a lot of `sm_buf_info`'s and `mp_obj_t` buffer objects hanging around, which I'm hoping to improve, but it seems to work.  In the saved file I can see

Screenshot from 2024-10-24 15-17-36

The four-line sections starting with `F7` and `F6` are written from alternating buffers, and have the expected content.

**Comment by timchinowsky** [comment]
Background playback of 8-channel audio from SD card to a 4-codec TAC5212 setup now working!  Details at https://github.com/timchinowsky/tac5.

**Comment by relic-se** [comment]
Hey @timchinowsky ! I've been loving this new feature and have recently implemented in a simple general purpose I2S library: https://github.com/relic-se/CircuitPython_I2SInOut. So far so good, but if you need any assistance with reviewing these updates further, let me know how I can help.

**Comment by timchinowsky** [comment]
> Hey @timchinowsky ! I've been loving this new feature and have recently implemented in a simple general purpose I2S library: https://github.com/relic-se/CircuitPython_I2SInOut. So far so good, but if you need any assistance with reviewing these updates further, let me know how I can help.

Hey that's great, thanks for letting me know!

**Comment by relic-se** [comment]
I'm not certain exactly what your goals in the recent updates are, but rather than building bespoke audio DSP processes for the rp2pio module, it would likely be better to find a way to integrate it into the existing audio systems of CircuitPython using the audiosample API. That way the existing processing modules could be used with custom PIO systems (ie: `audiomixer`, `audiofilters`, `audiodelays`, + future additions).

Alternatively, a separate `audiobusio.I2Sin` class could be created with the sole purpose of being an audio source in these existing modules (see `audiocore.RawSample` as an example). I've been working on this particular addition on the side, but it's not yet ready to share.

**Comment by relic-se** [comment]
It's also possible to feed data from rp2pio into an `audiocore.RawSample` buffer when double-buffering is enabled using the `background_read` functionality. See the "effect" example in the library I shared earlier for a demonstration of this: https://github.com/relic-se/CircuitPython_I2SInOut/blob/master/examples/i2sinout_effect.py.

**Comment by timchinowsky** [comment]
> I'm not certain exactly what your goals in the recent updates are, but rather than building bespoke audio DSP processes for the rp2pio module, it would likely be better to find a way to integrate it into the existing audio systems of CircuitPython using the audiosample API.

Yes, totally.  I was thinking when I started this that I would be leaning more on `audiocore`, but its limitations (<=16-bit, <=2 channels) made it difficult to explore the full capabilities of the codec I am working with.  Now that I've got the codec working well with `rp2pio` I'm going to see what can be done as far as unification.  For example, suppose that an 8-channel ADC could present as 8 mono `rawsample`-like objects or 4 stereo objects, etc., and that an 8-channel DAC could present as 4 stereo `I2SOut` objects, and so forth.

That said, I also like the idea of enabling `ulab`-like procedural processing of signals, to complement `synthio`'s declarative style.  I first got into CircuitPython as a tool for teaching electrical engineering lab classes, and making the signal processing more explicit and transparent would be great for that application.

I'm also planning to continue filling out the API for the TAC5212 codec.  For instance, the part has built in biquad filters on both ADCs and DACs (up to 3 per channel) and optional loopback between ADC and DAC, so the part can be used to implement 6-stage biquad analog-in/analog out filtering without any microcontroller involvement at all.

**Comment by relic-se** [comment]
> Yes, totally. I was thinking when I started this that I would be leaning more on `audiocore`, but its limitations (<=16-bit, <=2 channels) made it difficult to explore the full capabilities of the codec I am working with. Now that I've got the codec working well with `rp2pio` I'm going to see what can be done as far as unification. For example, suppose that an 8-channel ADC could present as 8 mono `rawsample`-like objects or 4 stereo objects, etc., and that an 8-channel DAC could present as 4 stereo `I2SOut` objects, and so forth.

Okay, I see where you're coming from here. Multi-channel output is fairly niche, so I don't think it's at the forefront of development in those areas. Yet, I don't see why one could operate with multiple mono/stereo outputs simultaneously as you've described for the time being.

As I've been playing around more with the API, I do think it's possible to connect an object like `audiocore.RawSample` to a `StateMachine` input and `audioio.AudioOut` as an output. For multi-channels, a few parameters like `channel_spacing` (total number of channels), `channel_offset` (starting channel index), and `channel_count` (mono/stereo) could be used to iterate through the appropriate buffers.

> That said, I also like the idea of enabling `ulab`-like procedural processing of signals, to complement `synthio`'s declarative style. I first got into CircuitPython as a tool for teaching electrical engineering lab classes, and making the signal processing more explicit and transparent would be great for that application.

Currently, the new `process` method feels out of scope for `rp2pio`. I think it would make much more sense to provide a separate DSP library that acts like `ulab` that you could operate directly on `WriteableBuffer` objects as you've described.

I'd really like to see the new `background_read` additions be added to the core. I think it might be best to save these new additions for another PR rather combine them with this one.

**Comment by timchinowsky** [comment]
Not sure why the `espressif` builds are failing, any suggestions @relic-se @jepler?  Thanks!  Perhaps related, my submodule understanding is fuzzy, do I need to do something get rid of the git warnings 
```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   ports/espressif/esp-idf (new commits)
	modified:   ports/raspberrypi/sdk (new commits)
```

**Review by dhalbert — CHANGES_REQUESTED** [review]
There are three regressions in submodule versions. See the first three entries in "Files changed". There shouldn't be any submodules listed in there for this PR> You can update submodules locally and then push again. The submodules don't get updated automatically necessarily when you merge from upstream.

**Comment by timchinowsky** [comment]
> There are three regressions in submodule versions. See the first three entries in "Files changed". There shouldn't be any submodules listed in there for this PR> You can update submodules locally and then push again. The submodules don't get updated automatically necessarily when you merge from upstream.

Thanks @dhalbert!  I can update locally with `make fetch-all-submodules` but then how do I push the submodules?  The various advice I am seeing on stackoverflow is not giving me a warm feeling.

**Comment by dhalbert** [comment]
> I can update locally with `make fetch-all-submodules` but then how do I push the submodules?

You commit the changed commits for the submodules.

**Comment by timchinowsky** [comment]
> > I can update locally with `make fetch-all-submodules` but then how do I push the submodules?
> 
> You commit the changed commits for the submodules.

@dhalbert thanks, but we would have had to get into details to debug what I was seeing.  I was trying to do as you say, but things were not acting as expected.  For now, I short-cut the process by cloning a fresh copy of main and re-applying the changes.

Now it looks like the build is too large for some boards?

**Comment by dhalbert** [comment]
I update submodules using this alias (for convenience):
```
alias gitsubupdate='git submodule sync --quiet --recursive && git submodule update --init --filter=blob:none'
```
Yes, the Wiznet boards were very close to full. Let's turn off USB host on those two boards. Add this to `mpconfigboard.mk` for those two:
```
CIRCUITPY_USB_HOST = 0
```

File: ports/raspberrypi/bindings/rp2pio/StateMachine.c [hunk 6]
```diff
@@ -553,27 +553,29 @@ static void fill_buf_info(sm_buf_info *info, mp_obj_t obj, size_t *stride_in_byt
 }
 
 static mp_obj_t rp2pio_statemachine_background_write(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
-    enum { ARG_once, ARG_loop, ARG_swap };
+    enum { ARG_once, ARG_loop, ARG_loop2, ARG_swap };
     static const mp_arg_t allowed_args[] = {
         { MP_QSTR_once,     MP_ARG_OBJ, {.u_obj = mp_const_none} },
         { MP_QSTR_loop,     MP_ARG_OBJ | MP_ARG_KW_ONLY, {.u_obj = mp_const_none} },
-        { MP_QSTR_swap,   MP_ARG_KW_ONLY | MP_ARG_BOOL, {.u_bool = false} },
+        { MP_QSTR_loop2,    MP_ARG_OBJ | MP_ARG_KW_ONLY, {.u_obj = mp_const_none} },
```
Comment (jepler) [inline ports/raspberrypi/bindings/rp2pio/StateMachine.c:560]: Please add loop2 to the documentation above.

File: ports/raspberrypi/bindings/rp2pio/StateMachine.c [hunk 7]
```diff
@@ -613,18 +616,145 @@ MP_DEFINE_CONST_FUN_OBJ_1(rp2pio_statemachine_get_writing_obj, rp2pio_statemachi
...
+//|         """
+//|         ...
+
+
+static mp_obj_t rp2pio_statemachine_background_read(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_once, ARG_loop, ARG_loop2, ARG_swap };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_once,     MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        { MP_QSTR_loop,     MP_ARG_OBJ | MP_ARG_KW_ONLY, {.u_obj = mp_const_none} },
+        { MP_QSTR_loop2,    MP_ARG_OBJ | MP_ARG_KW_ONLY, {.u_obj = mp_const_none} },
```
Comment (jepler) [inline ports/raspberrypi/bindings/rp2pio/StateMachine.c:685]: Please document loop2 in background_read

File: ports/raspberrypi/common-hal/rp2pio/StateMachine.c [hunk 8]
```diff
@@ -1036,123 +1060,331 @@ uint8_t rp2pio_statemachine_program_offset(rp2pio_statemachine_obj_t *self) {
...
+    } else {
+        if (!(self->loop_read_buf_info.info.len)) {
+            // If once_read_buf has nonzero length and loop_read_buf has zero length, read once_read_buf, loop2_read_buf and repeat last buf forever
+            self->loop_read_buf_info = self->loop2_read_buf_info;
+        } else {
+            if (!(self->loop2_read_buf_info.info.len)) {
+                // If once_read_buf has nonzero length and loop2_read_buf have zero length, read once_read_buf, loop_read_buf and repeat last buf forever
+                self->loop2_read_buf_info = self->loop_read_buf_info;
+
+            }
```
Comment (jepler) [inline ports/raspberrypi/common-hal/rp2pio/StateMachine.c:1265]: it feels like the handling of "once buf and two loop bufs" could be factored out into shared code. However, I'm not asking you to do this right now.

File: ports/raspberrypi/common-hal/rp2pio/StateMachine.c [hunk 9]
```diff
@@ -1036,123 +1060,331 @@ uint8_t rp2pio_statemachine_program_offset(rp2pio_statemachine_obj_t *self) {
...
+        self->once_read_buf_info.info.buf,
+        rx_source,
+        self->once_read_buf_info.info.len / stride_in_bytes,
+        false);
+
+    common_hal_mcu_disable_interrupts();
+    // Acknowledge any previous pending interrupt
+    dma_hw->ints0 |= 1u << channel_read;
+    MP_STATE_PORT(background_pio)[channel_read] = self;
+    dma_hw->inte1 |= 1u << channel_read;
```
Comment (jepler) [inline ports/raspberrypi/common-hal/rp2pio/StateMachine.c:1341]: should the assignment to `ints` be `ints1` to match `inte1`?

**Review by jepler — CHANGES_REQUESTED** [review]
I have some review comments. Thanks you in advance for addressing them!

File: ports/raspberrypi/common-hal/rp2pio/StateMachine.c (hunk 9 above)
Comment (timchinowsky) [reply ports/raspberrypi/common-hal/rp2pio/StateMachine.c:1341]: > should the assignment to `ints` be `ints1` to match `inte1`?

No doubt.  Thanks for the review @jepler!

**Comment by dhalbert** [comment]
There is trailing whitespace on line 148 of StateMachine.c

**Review by jepler — APPROVED** [review]
thanks, that addressed my requested changes.

**Review by dhalbert — APPROVED** [review]
submodule skew is fixed

**Comment by jepler** [comment]
I did a quick test of adafruit_neopxl8 which uses the old single buffered background write and it still functioned properly with Adafruit CircuitPython 9.2.0-38-g8d561a0b0e-dirty (this PR)

**Comment by timchinowsky** [comment]
Thanks for the test @jepler!

**Comment by relic-se** [comment]
I'm very excited to see this in 9.2.1. Good job everyone! 🎉

**Pull request merged**